Support building with GHC 9.2

RyanGlScott opened the pull request:
In GHC 9.2, enabling `UndecidableInstances` no longer implies enabling `FlexibleContexts` (see [here](https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/9.2?version_id=7e2ce63ba042c1934654c4316dc02028d8d3dd31#undecidableinstances-no-longer-implies-flexiblecontexts-in-instance-declarations)). As a result, I had to enable `FlexibleContexts` in `Language.Go.Parser`.

RyanGlScott committed:
Support building with GHC 9.2

In GHC 9.2, enabling `UndecidableInstances` no longer implies
enabling `FlexibleContexts` (see
[here](https://gitlab.haskell.org/ghc/ghc/-/wikis/migration/9.2?version_id=7e2ce63ba042c1934654c4316dc02028d8d3dd31#undecidableinstances-no-longer-implies-flexiblecontexts-in-instance-declarations)).
As a result, I had to enable `FlexibleContexts` in
`Language.Go.Parser`.

travitch commented on cabal.project:
```diff
@@ -0,0 +1 @@
+packages: .
```
What do you think about the `cabal.project.dist` pattern? I kind of like that for libraries (compared to top-level executables) because if you are in `crucible/deps/golang` making updates, your build commands will be able to build `crucible-go` without you having to change directories (because `cabal` won't pick up the spurious `cabal.project` file

RyanGlScott replied:
Good point. I've renamed this to `cabal.project.dist`, and I've added `cabal.project` to `.gitignore` to prevent `ln -s cabal.project.dist cabal.project` from dirtying the `git` tree.

RyanGlScott committed:
Add cabal.project.dist file

While not strictly needed for a single-project repo like `golang`, having a
`cabal.project.dist` file can be useful for temporarily making configuration
changes (e.g., using `head.hackage`). I've opted to make it a
`cabal.project.dist` file so that it does not interfere with builds by default
within in the context of a submodule. One can opt into the file by using
`ln -s cabal.project.dist cabal.project`.
